hashes/src/lib.rs

```diff
@@ -172,8 +171,7 @@ pub trait Hash:
 
     /// The byte array that represents the hash internally.
     type Bytes:
```

Comment:
OT but how come `Bytes` doesn't require `AsRef<[u8]>`?

Reply:
hmm, we have `borrow::Borrow<[u8]>` as a trait bound on `Hash`, I can't remember what discussion we had when we did `as_byte_array` and `from_byte_array` but was there a reason we did not have an `AsRef<[u8]>` trait bound on `Hash`? Either that or we should remove the `Borrow` bound, right?

Reply:
I think we decided that `Borrow` was "the right trait to express being byte-indexable". Our alternatives were `Deref` (which we all agree is bad magic) and `AsRef` (which TBH I don't understand how it differs from `Borrow`).

The `*_byte_array` methods though are separate; those let you get the underlying array storage as a complete owned object.

Reply:
I did read the "Relation to Borrow" [0] docs before posting but after reading your reply I read them again. I now think we are correct using `Borrow`, we should not bother having indexing trait bounds, and not use `AsRef`. I quickly hacked the repo to remove indexing bounds and everything works with just `borrow()` calls.

[0] https://doc.rust-lang.org/std/convert/trait.AsRef.html#relation-to-borrow

Reply:
> Yeah, I think let's just remove the bounds from the trait. And I'd also remove all Index bounds. Even PartialOrd might be too much.

I think I get this now, `PartialEq`, `Eq`, and `Ord` are needed for `Borrow` but `PartialOrd` is not, is that why you said this comment @Kixunil?

Reply:
> was there a reason we did not have an `AsRef<[u8]>` trait bound on `Hash`?

I can't think of a good reason.

> we should remove the `Borrow` bound, right?

Yes if we remove `Ord` because having the bound forces `Ord` to behave exactly as the array. No otherwise.

> which TBH I don't understand how it differs from `Borrow`

Different blanket impls and the additional requirement that `U: Borrow<T>` preserves `PartialEq`, `Ord`, `PartialOrd` ad `Hash` (homomorphism).

> we should not bother having indexing trait bounds,

I agree.

> and not use `AsRef`

Why? I think everything that can reasonably have `AsRef` should.